src/libraries/System.Private.CoreLib/src/System/Number.Dragon4.cs

```diff
@@ -100,203 +101,123 @@ public static unsafe void Dragon4Single(float value, int cutoffNumber, bool isSi
-                    BigInteger.SetUInt64(out scaledValue, 2 * mantissa);
-
-                    // scale           = 2 * 2^(-exponent)
-                    BigInteger.Pow2((uint)(-exponent + 1), out scale);
+            var len = descaledValue.GetLength();
 
-                    // scaledMarginLow = 2 * 2^(-1)
-                    BigInteger.SetUInt32(out scaledMarginLow, 1);
-                }
+            uint low = 0;
+            uint mid = 0;
+            uint high = 0;
 
-                // The high and low margins are equal
-                pScaledMarginHigh = &scaledMarginLow;
+            if (len > 0)
+            {
+                low = descaledValue.GetBlock(0);
             }
-
-            // Compute an estimate for digitExponent that will be correct or undershoot by one.
-            //
-            // This optimization is based on the paper "Printing Floating-Point Numbers Quickly and Accurately" by Burger and Dybvig http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.72.4656&rep=rep1&type=pdf
-            //
-            // We perform an additional subtraction of 0.69 to increase the frequency of a failed estimate because that lets us take a faster branch in the code.
-            // 0.69 is chosen because 0.69 + log10(2) is less than one by a reasonable epsilon that will account for any floating point error.
-            //
-            // We want to set digitExponent to floor(log10(v)) + 1
-            //      v = mantissa * 2^exponent
-            //      log2(v) = log2(mantissa) + exponent;
-            //      log10(v) = log2(v) * log10(2)
-            //      floor(log2(v)) = mantissaHighBitIdx + exponent;
-            //      log10(v) - log10(2) < (mantissaHighBitIdx + exponent) * log10(2) <= log10(v)
-            //      log10(v) < (mantissaHighBitIdx + exponent) * log10(2) + log10(2) <= log10(v) + log10(2)
-            //      floor(log10(v)) < ceil((mantissaHighBitIdx + exponent) * log10(2)) <= floor(log10(v)) + 1
-            const double Log10V2 = 0.30102999566398119521373889472449;
-            int digitExponent = (int)(Math.Ceiling(((int)(mantissaHighBitIdx) + exponent) * Log10V2 - 0.69));
-
-            // Divide value by 10^digitExponent.
-            if (digitExponent > 0)
+            if (len > 1)
             {
-                // The exponent is positive creating a division so we multiply up the scale.
-                scale.MultiplyPow10((uint)(digitExponent));
+                mid = descaledValue.GetBlock(1);
             }
-            else if (digitExponent < 0)
+            if (len > 2)
             {
-                // The exponent is negative creating a multiplication so we multiply up the scaledValue, scaledMarginLow and scaledMarginHigh.
+                high = descaledValue.GetBlock(2);
+            }
+            if (len > 3)
+            {
+                uint idk = descaledValue.GetBlock(3);
+                throw new Exception("TODO handle this somehow" + idk);
```

Comment:
If the above logic is correct, we theoretically should never reach this point, as state.scaledValue should be representing a valid decimal mantissa (and thus can fit in 3 uints). Unfortunately, some tests are hitting this exception.